### ISSUE TITLE
ssl_adapter.d(141): loading openssl: unknown version for init

### ISSUE THREAD

**Issue by GallaFrancesco** [ikod/dlang-requests#96]
```
object.Exception@/home/fra/.dub/packages/requests-1.0.6/requests/source/requests/ssl_adapter.d(141): loading openssl: unknown version for init
----------------
/home/fra/.dub/packages/requests-1.0.6/requests/source/requests/ssl_adapter.d:141 void requests.ssl_adapter._staticCtor_L50_C1() [0x77853ff]
??:? void requests.ssl_adapter.__modctor() [0x7785440]
??:? void rt.minfo.runModuleFuncs!(rt.minfo.ModuleGroup.runTlsCtors().__lambda1).runModuleFuncs(const(immutable(object.ModuleInfo)*)[]) [0xcaee4ac1]
??:? void rt.minfo.ModuleGroup.runTlsCtors() [0xcaee45f4]
??:? int rt.minfo.rt_moduleTlsCtor().__foreachbody1(ref rt.sections_elf_shared.DSO) [0xcaee4903]
??:? int rt.sections_elf_shared.DSO.opApply(scope int delegate(ref rt.sections_elf_shared.DSO)) [0xcaee5ff2]
??:? rt_moduleTlsCtor [0xcaee48ec]
??:? rt_init [0xcaedac6f]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() [0xcaedb13b]
??:? void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0xcaedb0d7]
??:? _d_run_main [0xcaedb044]
??:? main [0x77360d9]
??:? __libc_start_main [0xca74b4fa]
Program exited with code -4
```

ssl_adapter.d:141, gdb:
``` 
(gdb) info locals
libsslname = {"libssl.so", "libssl.so.1.1", "libssl.so.1.0.2", "libssl.so.1.0.1", "libssl.so.1.0.0"}
__r4343 = {"libssl.so", "libssl.so.1.1", "libssl.so.1.0.2", "libssl.so.1.0.1", "libssl.so.1.0.0"}
__key4344 = 0
lib = "libssl.so"
libcryptoname = {"libcrypto.so", "libcrypto.so.1.1", "libcrypto.so.1.0.2", "libcrypto.so.1.0.1", "libcrypto.so.1.0.0"}
__r4345 = {"libcrypto.so", "libcrypto.so.1.1", "libcrypto.so.1.0.2", "libcrypto.so.1.0.1", "libcrypto.so.1.0.0"}
__key4346 = 0
lib = "libcrypto.so"
init_matrix = {ptr = 0x7ffff6f1c000}
__aakey4352 = {__expand_field_0 = 1, __expand_field_1 = 0}
__aaval4353 = {ctxptr = 0x555555aca860 <_D8requests11ssl_adapter7opensslySQBgQBa7OpenSSL>, 
  funcptr = 0x5555558da504 <requests.ssl_adapter.OpenSSL.init1_0() const>}
__aakey4354 = {__expand_field_0 = 1, __expand_field_1 = 1}
__aaval4355 = {ctxptr = 0x555555aca860 <_D8requests11ssl_adapter7opensslySQBgQBa7OpenSSL>, 
  funcptr = 0x5555558da53c <requests.ssl_adapter.OpenSSL.init1_1() const>}
init = {ctxptr = 0x0, funcptr = 0x0}
```

**Comment by ikod** [comment]
Hello, @GallaFrancesco 

this problem means that library was unable to determine libssl version.
Can you please give me more info on your system and libssl installed?

**Comment by GallaFrancesco** [comment]
I'm running Gentoo linux with libressl-2.8.3 installed.

**Comment by ikod** [comment]
Never tested libressl. I can try to install virtual machine with gentoo and libressl to reproduce problem(it will take some time), but as a quick check - can you list ressl *.so files? it can be something like libresslX.Y.Y.so ?
Then we can try to load it instead of libssl here https://github.com/ikod/dlang-requests/blob/master/source/requests/ssl_adapter.d#L59

**Comment by GallaFrancesco** [comment]
```
$ ls /usr/lib64/libssl.*
/usr/lib64/libssl.so
/usr/lib64/libssl.so.46
/usr/lib64/libssl.so.46.0.1

$ ls /usr/lib64/libcrypto.so*
/usr/lib64/libcrypto.so
/usr/lib64/libcrypto.so.44
/usr/lib64/libcrypto.so.44.0.1

$ whereis libssl
libssl: /usr/lib64/libssl.so /usr/lib64/libssl3.so
```

It seems to me that the library is recognized (libssl.so)

**Comment by ikod** [comment]
well, that means that libressl API differs in some unexpected (for me) way. I'll install gentoo with libressl and check what goes wrong

**Comment by GallaFrancesco** [comment]
I think I found the issue. From the manpage of libressl, `man  `:

```
          OPENSSL_VERSION_NUMBER:
           0x000906000 == 0.9.6 dev
           0x000906023 == 0.9.6b beta 3
           0x00090605f == 0.9.6e release
           0x020000000 == 2.0.0 for any version of LibreSSL

           LIBRESSL_VERSION_NUMBER:
           0x02070000f == LibreSSL 2.7.0
```

`Openssl_version_detect` in ssl_adapter.d:193, gdb:
```
(gdb) p v
$1 = 536870912
```
Which once bit-shifted results in `Version(0, 2)`, which doesn't seem to be correct. 
The manpage also specifies that libressl method `OpenSSL_version_num` is available since libressl 2.7.1.

**Comment by FraMecca** [comment]
Can confirm that gentoo with libressl 2.6.5 works.

**Comment by ikod** [comment]
Great! Anyway I will test libressl detection and commit fix probably at weekend.

**Comment by ikod** [comment]
resolved with issue #99
